.. _references:

References
==========

**Bibliography**

Briggs, J.A., Weinreb, C., Wagner, D.E., Megason, S., Peshkin, L., Kirschner, M.W. and Klein, A.M. (2018).
**The dynamics of gene expression in vertebrate embryogenesis at single-cell resolution.** *Science*, **360(6392)**. [https://doi.org/10.1126/science.aar5780](https://doi.org/10.1126/science.aar5780)

Cazet, J., Siebert, S., Little, H.M., Bertemes, P., Primack, A.S., Ladurner, P., Achrainer, M., Fredriksen,
M.T., Moreland, R.T., Singh, S. and Zhang, S. (2022). **New Hydra genomes reveal conserved principles of hydrozoan
transcriptional regulation.** bioRxiv. [https://doi.org/10.1101/2022.06.21.496857](https://doi.org/10.1101/2022.06.21.496857)

Cao, J., Spielmann, M., Qiu, X., Huang, X., Ibrahim, D.M., Hill, A.J., Zhang, F., Mundlos, S., Christiansen, L.,
Steemers, F.J. and Trapnell, C. (2019). **The single-cell transcriptional landscape of mammalian organogenesis.**
*Nature*, **566(7745)**. [https://doi.org/10.1038/s41586-019-0969-x](https://doi.org/10.1038/s41586-019-0969-x)

Cheng, S., Pei, Y., He, L., Peng, G., Reinius, B., Tam, P.P., Jing, N. and Deng, Q. (2019). **Single-cell RNA-seq
reveals cellular heterogeneity of pluripotency transition and X chromosome dynamics during early mouse development.**
*Cell reports*, **26(10)**. [https://doi.org/10.1016/j.celrep.2019.02.031](https://doi.org/10.1016/j.celrep.2019.02.031)

Domazet-Loso, T., Brajkovic J. and Tautz D. (2007). **A phylostratigraphy approach to uncover the genomic history of
major adaptations in metazoan lineages.** *Trends in Genetics*, **23(11)**. [https://doi.org/10.1016/j.tig.2007.08.014](https://doi.org/10.1016/j.tig.2007.08.014)

Domazet-Loso, T. and Tautz. D. (2010). **A phylogenetically based transcriptome age index mirrors ontogenetic
divergence patterns.** *Nature*, **468(7325)**. [https://doi.org/10.1038/nature09632](https://doi.org/10.1038/nature09632)

Drost, H.G., Gabel, A., Grosse, I. and Quint M. (2015). **Evidence for active maintenance of phylotranscriptomic
hourglass patterns in animal and plant embryogenesis.** *Molecular biology and evolution*, **32(5)**. [https://doi.org/10.1093%2Fmolbev%2Fmsv012](https://doi.org/10.1093%2Fmolbev%2Fmsv012)

Drost, H.G., Gabel, A., Liu, J., Quint, M. and Grosse, I. (2018). **myTAI: evolutionary transcriptomics with R.**
*Bioinformatics*, **34(9)**. [https://doi.org/10.1093/bioinformatics/btx835](https://doi.org/10.1093/bioinformatics/btx835)

Emms, D.M. and Kelly, S. (2019). **OrthoFinder: phylogenetic orthology inference for comparative genomics.** 
*Genome biology*, **20(1)**. [https://doi.org/10.1186/s13059-019-1832-y](https://doi.org/10.1186/s13059-019-1832-y)

Farrell, J.A., Wang, Y., Riesenfeld, S.J., Shekhar, K., Regev, A. and Schier, A.F. (2018). **Single-cell reconstruction
of developmental tra jectories during zebrafish embryogenesis.** *Science*, **360(6392)**. [https://doi.org/10.1126/science.aar3131](https://doi.org/10.1126/science.aar3131)

Gossmann, T.I., Saleh, D., Schmid, M.W., Spence, M.A. and Schmid, K.J. (2016). **Transcriptomes of plant gametophytes
have a higher proportion of rapidly evolving and young genes than sporophytes.** *Molecular biology and evolution*,
**33(7)**. [https://doi.org/10.1093/molbev/msw044](https://doi.org/10.1093/molbev/msw044)

Harris, C.R., Millman, K.J., Van Der Walt, S.J., Gommers, R., Virtanen, P., Cournapeau, D., Wieser, E., Taylor, J.,
Berg, S., Smith, N.J. and Kern, R. (2020). **Array programming with NumPy.** *Nature*, **585(7825)**. [https://doi.org/10.1038/s41586-020-2649-2](https://doi.org/10.1038/s41586-020-2649-2)

Hernández-Plaza, A., Szklarczyk, D., Botas, J., Cantalapiedra, C.P., Giner-Lamia, J., Mende, D.R., Kirsch, R.,
Rattei, T., Letunic, I., Jensen, L.J. and Bork, P. (2022). **eggNOG 6.0: enabling comparative genomics across
12 535 organisms.** *Nucleic Acids Research*, **51(D1)**. [https://doi.org/10.1093%2Fnar%2Fgkac1022](https://doi.org/10.1093%2Fnar%2Fgkac1022)

Huerta-Cepas, J., Serra, F. and Bork, P. (2016). **ETE 3: reconstruction, analysis, and visualization of phylogenomic data.**
*Molecular biology and evolution*, **33(6)**. [https://doi.org/10.1093/molbev/msw046](https://doi.org/10.1093/molbev/msw046)

Julca, I., Ferrari, C., Flores-Tornero, M., Proost, S., Lindner, A.-C., Hackenberg, D., Steinbachov́a, L.,
Michaelidis, C., Gomes Pereira, S., Misra, C.S., Kawashima, T., Borg, M., Berger, F., Goldberg, J., Johnson, M.,
Honys, D., Twell, D., Sprunck, S., Dresselhaus, T., Becker, J.D., Mutwil, M. (2021).
**Comparative transcriptomic analysis reveals conserved programmes underpinning organogenesis and reproduction in
land plants.** *Nature Plants*, **7(8)**. [https://doi.org/10.1038/s41477-021-00958-2](https://doi.org/10.1038/s41477-021-00958-2)

Kiełbasa S.M., Wan R., Sato K., Horton P., Frith M.C. (2011). **Adaptive seeds tame genomic sequence comparison.**
*Genome Research*, *21(3)**. [https://doi.org/10.1101/gr.113985.110](https://doi.org/10.1101/gr.113985.110)

Liu, J. and Robinson-Rechavi, M. (2018). **Developmental Constraints on Genome Evolution in Four Bilaterian Model Species**
*Genome Biology and Evolution*, **10(9)**. [https://doi.org/10.1093/gbe/evy177](https://doi.org/10.1093/gbe/evy177)

Liu, J., Frochaux, M. , Gardeux, V., Deplancke, B. and Robinson-Rechavi, M. (2020). **Inter-embryo gene expression
variability recapitulates the hourglass pattern of evo-devo.** *BMC biology*, **18(1)**. [https://doi.org/10.1186/s12915-020-00842-z](https://doi.org/10.1186/s12915-020-00842-z)

Ma, F., Lau, C.Y. and Zheng, C. (2021). **Large genetic diversity and strong positive selection in F-box and GPCR genes
among the wild isolates of Caenorhabditis elegans.** *Genome Biology and Evolution*, **13(5)**. [https://doi.org/10.1093/gbe/evab048](https://doi.org/10.1093/gbe/evab048)

Ma, F. and Zheng, C. (2023). **Transcriptome age of individual cell types in caenorhabditis elegans.**
*Proceedings of the National Academy of Sciences*, **120(9)**. [https://doi.org/10.1073/pnas.2216351120](https://doi.org/10.1073/pnas.2216351120)

McInnes, L., Healy, J. and Melville, J. (2018). **Umap: Uniform manifold approximation and projection for dimension
reduction.** *arXiv preprint*, **arXiv:1802.03426**. [https://doi.org/10.48550/arXiv.1802.03426](https://doi.org/10.48550/arXiv.1802.03426)

McKinney, W. (2011). **pandas: a foundational python library for data analysis and statistics.**
*Python for high performance and scientific computing*, **14(9)**. [https://www.dlr.de/sc/portaldata/15/resources/dokumente/pyhpc2011/submissions/pyhpc2011_submission_9.pdf](https://www.dlr.de/sc/portaldata/15/resources/dokumente/pyhpc2011/submissions/pyhpc2011_submission_9.pdf)

Mohammed, H., Hernando-Herraez, I., Savino, A., Scialdone, A., Macaulay, I., Mulas, C., Chandra, T., Voet, T.,
Dean, W., Nichols, J. and Marioni, J.C. (2017). **Single-cell landscape of transcriptional heterogeneity and cell fate
decisions during mouse early gastrulation.** *Cell reports*, **20(5)**. [https://doi.org/10.1016/j.celrep.2017.07.009](https://doi.org/10.1016/j.celrep.2017.07.009)

Packer, J.S., Zhu, Q., Huynh, C., Sivaramakrishnan, P., Preston, E., Dueck, H., Stefanik, D.,
Tan, K., Trapnell, C., Kim, J. and Waterston, R.H., (2019).
**A lineage-resolved molecular atlas of C. elegans embryogenesis at single-cell resolution.**
*Science*, **365(6459)**. [https://doi.org/10.1126/science.aax1971](https://doi.org/10.1126/science.aax1971)

Pijuan-Sala, B., Griffiths, J.A., Guibentif, C., Hiscock, T.W., Jawaid, W., Calero-Nieto, F.J., Mulas, C.,
Ibarra-Soria, X., Tyser, R.C., Ho, D.L.L. and Reik, W. (2019). **A single-cell molecular map of mouse gastrulation and
early organogenesis.** *Nature*, **566(7745)**. [https://doi.org/10.1038/s41586-019-0933-9](https://doi.org/10.1038/s41586-019-0933-9)

Qiu, C., Cao, J., Martin, B.K., Li, T., Welsh, I.C., Srivatsan, S., Huang, X., Calderon,
D., Noble, W.S., Disteche, C.M. and Murray, S.A. (2022). 
**Systematic reconstruction of cellular trajectories across mouse embryogenesis.** 
*Nature genetics*, **54(3)**. [https://doi.org/10.1038/s41588-022-01018-x](https://doi.org/10.1038/s41588-022-01018-x)

Quint, M., Drost, H.-G., Gabel, A., Ullrich, K.K., Bönn, M. and Grosse, I. (2012). **A transcriptomic hourglass in
plant embryogenesis.** *Nature*, **490(7418)**. [https://doi.org/10.1038/nature11394](https://doi.org/10.1038/nature11394)

Sun, S., Rödelsperger, C. and Sommer, R.J., (2021).
**Single worm transcriptomics identifies a developmental core network of oscillating genes with deep conservation across nematodes.**
*Genome research*, **31(9)**. [https://doi.org/10.1101/gr.275303.121](https://doi.org/10.1101/gr.275303.121)

Tarashansky, A.J., Musser, J.M., Khariton, M., Li, P., Arendt, D., Quake, S.R. and Wang, B. (2021).
**Mapping single-cell atlases throughout metazoa unravels cell type evolution.** *Elife*, **10**. [https://doi.org/10.7554/elife.66747](https://doi.org/10.7554/elife.66747)

Ullrich, K.K. (2020). **CRBHits: From conditional reciprocal best hits to codon alignments and Ka/Ks in R.**
*The Journal of Open Source Software*, **5(55)**. [https://doi.org/10.21105/joss.02424](https://doi.org/10.21105/joss.02424)

Van Bel, M., Silvestri, F., Weitz, E.M., Kreft, L., Botzki, A., Coppens, F. and Vandepoele, K. (2022).
**PLAZA 5.0: extending the scope and power of comparative and functional genomics in plants.**
*Nucleic Acids Research*, **50(D1)**. [https://doi.org/10.1093/nar/gkab1024](https://doi.org/10.1093/nar/gkab1024)

Wagner, D.E., Weinreb, C., Collins, Z.M., Briggs, J.A., Megason, S.G. and Klein, A.M. (2018).
**Single-cell mapping of gene expression landscapes and lineage in the zebrafish embryo.** *Science*, **360(6392)**. [https://doi.org/10.1126%2Fscience.aar4362](https://doi.org/10.1126%2Fscience.aar4362)

Wolf, F.A., Angerer, P. and Theis, F.J. (2018). **SCANPY: large-scale single-cell gene expression data analysis.**
*Genome biology*, **19(1)**. [https://doi.org/10.1186/s13059-017-1382-0](https://doi.org/10.1186/s13059-017-1382-0)
